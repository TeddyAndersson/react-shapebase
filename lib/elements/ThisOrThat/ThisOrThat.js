import React from 'react';
import _ from 'lodash';
import styled from 'styled-components'
import PropTypes from 'prop-types';
import RectangleShape from '../Shape/RectangleShape'

const arrayOfLength = function(expectedLength, props, propName, componentName) {
  const arrayPropLength = props[propName].length

  if (arrayPropLength !== expectedLength) {
    return new Error(
      `Invalid array length ${arrayPropLength} (expected ${expectedLength}) for prop ${propName} supplied to ${componentName}. Validation failed.`
    );
  }
}

class Element extends React.Component{
  constructor(props){
    super(props);
    this.state = {
      activeOption: this.props.initialOption,
    }
  }

  handleClick = (id) => {
    this.setState({activeOption: id})

    _.invoke(this.props, 'onClick', id, this.props)
  }

  render(){
      const {
        className,
        activeBackground,
      } = this.props
    return (
        <ul className={className}>
          { _.values(this.props.options).map(option => (
            // Replace <li/> with a component for better state handling and generic code
            <li
                key={option.id}
                onClick={() => this.handleClick( option.id )}
            >
              <RectangleShape
                background={`${ option.id === this.state.activeOption ? activeBackground : 'rgba(0,0,0,0.25)'}`}
                raised={option.id === this.state.activeOption ? 1: 0}
              >
                <input
                  type="hidden"
                  name={option.name}
                  value={option.value}
                  autoComplete="off"
                />
                <div>
                  <p>
                    <span>{option.label}</span>
                  </p>
                </div>
              </RectangleShape>
            </li>
        ))}
    </ul>
    )
  };
};

Element.propTypes = {
  className: PropTypes.string,
  options: arrayOfLength.bind(null, 2),
  onClick: PropTypes.func.isRequired,
  /**
   * Property to set the inital option.
   */
  initialOption: PropTypes.string,
  /**
   * Property to change the background for an active option.
   */
  activeBackground: PropTypes.string,
};

Element.defaultProps = {
  initialOption: undefined,
  activeBackground: '#1FAACB',
}

const ThisOrThat = styled(Element)`
    padding: 0;
    list-style: none;
    display: flex;
    margin: 0;
    justify-content: space-between;
    color: #fff;

    li {
        width: 48%;

        p {
            font-weight: bolder;
            text-align: center;
            margin: 0;
            font-size: 1em;
        }

        div {
            width: auto;
            padding: 0.1em 0.8em;
            max-height: 3em;
            display: flex;
            justify-content: center;
            align-items: center;
            transition: all 0.3s ease
        }
    }
`;

export default ThisOrThat;
