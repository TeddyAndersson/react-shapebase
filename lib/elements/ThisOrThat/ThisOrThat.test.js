import React from 'react';
import ThisOrThat from './ThisOrThat'
import renderer from 'react-test-renderer';

const lorem = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."

test('should render ThisOrThat', () => {
  const component = renderer.create(
    <ThisOrThat
      onClick={()=>{}}
      value=""
      options={[
        {
          id: '1',
          name: 'T-rex',
          value: '1',
          label: 'T-rex',
          icon: 'mars',
          background: 'burlywood',
        },
        {
          id: '2',
          name: 'Raptor',
          value: '2',
          label: 'Raptor',
          icon: 'transgender',
          background: 'silver',
        },
      ]}
    />

  );
  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();
})
