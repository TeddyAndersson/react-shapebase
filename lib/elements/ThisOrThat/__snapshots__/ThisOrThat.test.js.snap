// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ThisOrThat 1`] = `
<ul
  className="sc-htpNat leYLMy"
>
  <li
    onClick={[Function]}
  >
    <div
      className="sc-bwzfXH lipQua sc-bdVaJa jmjnGv"
    >
      <input
        autoComplete="off"
        name="T-rex"
        type="hidden"
        value="1"
      />
      <div>
        <p>
          <span>
            T-rex
          </span>
        </p>
      </div>
    </div>
  </li>
  <li
    onClick={[Function]}
  >
    <div
      className="sc-bwzfXH lipQua sc-bdVaJa jmjnGv"
    >
      <input
        autoComplete="off"
        name="Raptor"
        type="hidden"
        value="2"
      />
      <div>
        <p>
          <span>
            Raptor
          </span>
        </p>
      </div>
    </div>
  </li>
</ul>
`;
